### PR TITLE
Fix Handling of HTTP CONNECT Header in Proxy Connections

### PR DESCRIPTION
**Pull request by bearx3f**
From #7651 After I tested the latest pull request, which I'd previously said passed, the proxy server still couldn't sign the certificate when deployed with the actual application. This was because the server didn't receive the remote server in the correct `domain:port` format, such as `example.com:443`. In this patch, I've fixed the `CONNECT` request to be correct.

I noticed this by observing the request sent by Curl, which sends:

```
CONNECT example.com:443 HTTP/1.0
Host: example.com:443
```

Originally, Libtorrent was sending:

```
CONNECT 23.220.75.232:443 HTTP/1.0
Host: example.com
```

In this original request, the Host header was missing the port, and the **CONNECT** method was sending an IP address instead of a domain name.

### CODE REVIEW

File: include/libtorrent/http_stream.hpp [hunk 1]
```diff
@@ -98,6 +99,80 @@ class http_stream : public proxy_base
...
+	// - if port == 0, return host unchanged
+	// - if host is bracketed IPv6: [addr] or [addr]:port, append port only if missing
+	// - if host contains no ':', append ":port"
+	// - if host contains colon(s):
+	//     * if suffix after last ':' is all digits, assume it's already a host:port -> leave unchanged
+	//     * otherwise treat it as an (unbracketed) IPv6 literal or hostname with colons and wrap
+	//       it in brackets and append :port -> [host]:port
+	static std::string format_host_for_connect(std::string host, unsigned short const port)
+	{
+		if (port == 0) return host;
```
Comment (arvidn) [inline include/libtorrent/http_stream.hpp]:
```suggestion
		if (port == 80) return host;
```
I imagine this is what you meant, right?

File: include/libtorrent/http_stream.hpp [hunk 2]
```diff
@@ -98,6 +99,80 @@ class http_stream : public proxy_base
...
 
+	// format a hostname (not a numeric IP) with a port for an HTTP CONNECT request.
+	// rules:
+	// - if port == 0, return host unchanged
+	// - if host is bracketed IPv6: [addr] or [addr]:port, append port only if missing
+	// - if host contains no ':', append ":port"
+	// - if host contains colon(s):
+	//     * if suffix after last ':' is all digits, assume it's already a host:port -> leave unchanged
+	//     * otherwise treat it as an (unbracketed) IPv6 literal or hostname with colons and wrap
+	//       it in brackets and append :port -> [host]:port
```
Comment (arvidn) [inline include/libtorrent/http_stream.hpp]: rather than describing exactly what this function is doing, in detail, it would be much better to describy *why* it's doing it. Perhaps even with a reference to the appropriate RFC describing the host header format. And maybe with a quote from it

File: include/libtorrent/http_stream.hpp [hunk 3]
```diff
@@ -98,6 +99,80 @@ class http_stream : public proxy_base
...
+	static std::string format_host_for_connect(std::string host, unsigned short const port)
+	{
+		if (port == 0) return host;
+
+		if (!host.empty() && host.front() == '[')
+		{
+			auto const rb = host.find(']');
+			bool const has_port = (rb != std::string::npos && rb + 1 < host.size() && host[rb + 1] == ':');
```
Comment (arvidn) [inline include/libtorrent/http_stream.hpp]: does the host sometimes come with a port? that doesn't sound right.
I think it should either be an assert that it doesn't have a port, or a comment explaining why it (sometimes) has a port

File: include/libtorrent/http_stream.hpp (hunk 1 above)
Comment (bearx3f) [reply include/libtorrent/http_stream.hpp]: In some case port=0 there's no port to be append

File: include/libtorrent/http_stream.hpp (hunk 3 above)
Comment (bearx3f) [reply include/libtorrent/http_stream.hpp]: I will assume that host will not already has port and then assert there's port in host

File: include/libtorrent/http_stream.hpp (hunk 2 above)
Comment (bearx3f) [reply include/libtorrent/http_stream.hpp]: I will change it

File: include/libtorrent/http_stream.hpp (hunk 3 above)
Comment (bearx3f) [reply include/libtorrent/http_stream.hpp]: and then add ASSERT `host`

File: include/libtorrent/http_stream.hpp [hunk 4]
```diff
@@ -98,6 +99,41 @@ class http_stream : public proxy_base
...
+	// Format a hostname with port for HTTP CONNECT request per RFC 9110 Section 9.3.6.
+	// The authority component must be in the form "host:port" where IPv6 literals
+	// are enclosed in square brackets as defined in RFC 3986 Section 3.2.2.
+	// Assumes host parameter contains no port suffix.
+	// Rules:
+	// - if port == 0, return host unchanged
+	// - if host is already bracketed IPv6: [addr], append ":port"
+	// - if host contains colons (IPv6), bracket and append ":port" -> [host]:port
+	// - otherwise append ":port" for regular hostnames/IPv4
+	static std::string format_host_for_connect(std::string host, unsigned short const port)
```
Comment (arvidn) [inline include/libtorrent/http_stream.hpp]: I think this should be a free function (not a static member). It should be defined in a `.cpp` file. And it should have a unit test, to make sure it handles all the edge cases correctly?
Since `http_stream` doesn't have a `.cpp` file, maybe it would make sense to put this in `string_util.cpp`, with the test in `test_string.cpp`

File: test/http_proxy.py [hunk 5]
```diff
@@ -209,15 +209,20 @@ def connect_request(self):
...
+                self.log_message("Host header matches CONNECT target")
+            else:
+                self.log_message("Host header differs from CONNECT target")
         else:
-            self.log_message("CONNECT request without Host header (target: %s:%s)",
+            self.log_message("CONNECT request: target=%s:%s, NO Host header", 
                            host, port)
 
         # If require_host_header is set, reject requests without Host header
-        if self.require_host_header and not host_header:
```
Comment (arvidn) [inline test/http_proxy.py:219]: you you remove the check for `self.require_host_header` (and make it mandatory), you should remove the comment above as well as *setting* this member to `True` at the top of the class

Comment (bearx3f) [reply test/http_proxy.py:219]: I'm really sorry.

**Comment by arvidn** [comment]
it looks like the simulation error is legitimate and needs to be addressed

**Comment by bearx3f** [comment]
> it looks like the simulation error is legitimate and needs to be addressed

The problem is my patch will always send Host header but simulation test check if send_host_in_connect=false host should be empty. I will fix.

File: include/libtorrent/http_stream.hpp [hunk 6]
```diff
@@ -128,10 +128,24 @@ class http_stream : public proxy_base
 		// send CONNECT
 		std::back_insert_iterator<std::vector<char>> p(m_buffer);
 		std::string const endpoint = print_endpoint(m_remote_endpoint);
-		write_string("CONNECT " + endpoint + " HTTP/1.0\r\n", p);
+		// if we were given the original host (domain or IP), prefer using it (lets proxy resolve domains)
 		if (!m_host.empty())
 		{
-			write_string("Host: " + m_host + "\r\n", p);
+			std::string const remote_host = ::libtorrent::format_host_for_connect(
+				std::move(m_host),
```
Comment (arvidn) [inline include/libtorrent/http_stream.hpp:135]: this `std::move()` looks really suspicious. It means this socket can't be reused, which may be an unexpected behavior. Are you certain that this is correct?

Since `m_host` is not passed in via `async_connect()`, I'm pretty sure you can end up setting the host only once, and the connect the socket multiple times, e.g. if the connection fails for some reason.

I don't think you should move `m_host` here.
```suggestion
				m_host, 
```

File: include/libtorrent/http_stream.hpp [hunk 7]
```diff
@@ -128,10 +128,24 @@ class http_stream : public proxy_base
 		// send CONNECT
 		std::back_insert_iterator<std::vector<char>> p(m_buffer);
 		std::string const endpoint = print_endpoint(m_remote_endpoint);
-		write_string("CONNECT " + endpoint + " HTTP/1.0\r\n", p);
+		// if we were given the original host (domain or IP), prefer using it (lets proxy resolve domains)
 		if (!m_host.empty())
 		{
-			write_string("Host: " + m_host + "\r\n", p);
+			std::string const remote_host = ::libtorrent::format_host_for_connect(
```
Comment (arvidn) [inline include/libtorrent/http_stream.hpp:134]:
```suggestion
			std::string const remote_host = format_host_for_connect(
```

**Comment by arvidn** [comment]
sorry for dropping the ball on this one. I think it looks good. just a few small tweaks

File: test/http_proxy.py [hunk 8]
```diff
@@ -152,7 +152,7 @@ class Handler(http.server.BaseHTTPRequestHandler):
     close_connection = True
     timeout = 30
     basic_auth = None
-    require_host_header = False  # Set to True to require Host header in CONNECT requests
+    require_host_header = True  # Always require Host header in CONNECT requests for proper HTTP compliance
```
Comment (arvidn) [inline test/http_proxy.py:155]: I believe I was thinking that this was part of the `BaseHTTPRequestHAndler`, but I don't think it is. Now you can probably just remove this field entirely

**Comment by arvidn** [comment]
@bearx3f how does this look? https://github.com/arvidn/libtorrent/pull/8155

**Comment by bearx3f** [comment]
Thank you, for you resopnse

I don't have much experience with C++, so I can't really provide feedback on your changes. Please feel free to use your best judgment. Also, just to mention, the code I submitted was written with the help of AI.
